Amend the lib/ code for easier inclusion in other programs

avar opened the pull request:
This introduces no functional changes, but allows the lib/ code to be
used as-is in other programs without patching these files directly.

With these changes the git project can use this code without any
modifications to the upstream code.

This is used by a patch series I've just submitted as an RFC against git.git: https://public-inbox.org/git/20170517113824.31700-1-avarab@gmail.com/T/#u

Currently that patch series points at my fork, but it would be great if you could merge this (or if you'd like to have this done differently, I'm happy to change it), then projects like git could use your code as-is via either a submodule, or just by copying the code as-is with no customizations to the files themselves, only to the Makefile that makes use of them.

shumow commented on README.md:
```diff
@@ -78,3 +78,16 @@ else
 
+## Inclusion in other programs
+
+This code can easily be included in other programs. Rather than
```
Please consider changing to:
"In order to make it easier to include these sources in other project there are several preprocessor macros that the code uses.  Rather than copy/pasting and customizing or specializing the code, first see if setting any of these defines appropriately will allow you to avoid modifying the code yourself."

avar replied:
Done in the patch I'm about to push. Thanks!

shumow commented on README.md:
```diff
@@ -78,3 +78,16 @@ else
+minimal modifications to your build system:
+
+- SHA1DC_NO_STANDARD_INCLUDES -- Skips including standard headers
+- SHA1DC_CUSTOM_INCLUDE_SHA1_C -- Includes a custom header at the top of sha1.c
```
How standard is the practice of using "custom" includes?  I have never used this approach before.  However, I am by no means an expansive contributor to C/C++ Open Source Projects.  While this is flexible, if it is not a commonly used approach it may be overlooked rather than used, defeating the purpose of the changes.

avar replied:
Answered below.

shumow commented on README.md:
```diff
@@ -78,3 +78,16 @@ else
+- SHA1DC_INIT_SAFE_HASH_DEFAULT -- Sets the default for safe_hash in SHA1DCInit()
+- SHA1DC_CUSTOM_TRAILING_INCLUDE_SHA1_C -- Includes a custom trailer in sha1.c
+- SHA1DC_CUSTOM_TRAILING_INCLUDE_SHA1_H -- Includes a custom trailer in sha1.h
+- SHA1DC_CUSTOM_INCLUDE_UBC_CHECK_C -- Includes a custom header at the top of ubc_check.c
```
See previous comments about defines that add custom include/header fiels.

avar replied:
Answered below.

shumow commented on README.md:
```diff
@@ -78,3 +78,16 @@ else
+of these defines doesn't allow you to use the code as-is with only
+minimal modifications to your build system:
+
+- SHA1DC_NO_STANDARD_INCLUDES -- Skips including standard headers
```
If this is set, than the other custom header must be set, correct?  Otherwise important includes/defines will not occur.  How common is it that projects do not use the standard includes.  Do you have a specific project that does not use the standard includes in mind?  If so, which one?

avar replied:
In general. All of this is needed so that the git project (as in, the git binary you have sitting on your computer) can use this code as-is without monkeypatching the code. These changes are all made in places where we are now injecting custom code. See:

* https://github.com/git/git/commit/45a574eec8
* https://github.com/git/git/commit/c0c20060af
* https://github.com/git/git/commit/8325e43b82

To answer this specific question. Yes it is common for highly portable code to want code it uses to not do standard header includes for whatever reason. E.g. this code will include stdint.h, but git wants to include either inttypes.h or stdint.h depending on a compatibility switch with some odd platforms.

shumow commented on lib/sha1.c:
```diff
@@ -5,10 +5,20 @@
 #include <memory.h>
 #include <stdio.h>
 #include <stdlib.h>
+#endif
```
Stylewise, I prefer to add a comment after a "#endif" showing which preprocessor macro is controlling the condition (or a comment about what the check was about.)  That said, it is not a requirement.

avar replied:
I don't mind changing this at all, but this does not appear to be the established coding style of this project:

    $ git grep --no-line-number -h '#endif' origin/master|sort|uniq -c|sort -nr
        253 #endif
          1 #endif /*ENDIANNESS SELECTION*/
          1 #endif /*BUILDNOCOLLDETECTSHA1COMPRESSION*/

The only ones with comments are much longer ifdef spans, which doesn't apply to any of the ones I'm adding.

shumow commented on README.md:
```diff
@@ -78,3 +78,16 @@ else
+
+- SHA1DC_NO_STANDARD_INCLUDES -- Skips including standard headers
+- SHA1DC_CUSTOM_INCLUDE_SHA1_C -- Includes a custom header at the top of sha1.c
+- SHA1DC_INIT_SAFE_HASH_DEFAULT -- Sets the default for safe_hash in SHA1DCInit()
```
I think we should add a comment that this define should be set to either 0 or 1.

avar replied:
Fixed in the push I'm about to make. Thanks.

shumow commented:
Ævar,

Thanks for your contributions here!  I made some comments, so please take a
look at those.  Also, I think we should wait for Marc to weigh in as well.

-Dan

On Wed, May 17, 2017 at 4:41 AM, Ævar Arnfjörð Bjarmason <
notifications@github.com> wrote:

> This introduces no functional changes, but allows the lib/ code to be
> used as-is in other programs without patching these files directly.
>
> With these changes the git project can use this code without any
> modifications to the upstream code.
>
> This is used by a patch series I've just submitted as an RFC against
> git.git: https://public-inbox.org/git/20170517113824.31700-1-avarab@
> gmail.com/T/#u
>
> Currently that patch series points at my fork, but it would be great if
> you could merge this (or if you'd like to have this done differently, I'm
> happy to change it), then projects like git could use your code as-is via
> either a submodule, or just by copying the code as-is with no
> customizations to the files themselves, only to the Makefile that makes use
> of them.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cr-marcstevens/sha1collisiondetection/pull/31
> Commit Summary
>
>    - Amend the lib/ code for easier inclusion in other programs
>
> File Changes
>
>    - *M* README.md
>    <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31/files#diff-0>
>    (13)
>    - *M* lib/sha1.c
>    <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31/files#diff-1>
>    (16)
>    - *M* lib/sha1.h
>    <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31/files#diff-2>
>    (6)
>    - *M* lib/ubc_check.c
>    <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31/files#diff-3>
>    (5)
>    - *M* lib/ubc_check.h
>    <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31/files#diff-4>
>    (2)
>
> Patch Links:
>
>    - https://github.com/cr-marcstevens/sha1collisiondetection/pull/
>    31.patch
>    - https://github.com/cr-marcstevens/sha1collisiondetection/pull/31.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cr-marcstevens/sha1collisiondetection/pull/31>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOhEpPrOc9zHAK1keNorHhk7mAkFshRQks5r6tzggaJpZM4NdvgR>
> .
>

avar commented:
@shumow: Thanks a lot for the detailed review. I've hopefully addressed all the comments above, and have just pushed a new patch. I also took the opportunity to fleshen out the commit message in much more detail.

The only thing I didn't take into your account was adding comments at the end of `#endif`. As noted I really don't care either way, but just from looking at the current code this looks like the opposite of the style Marc initially used, so I've used something that looks more like his style, so I though it was more likely that he'd reject this if I made that change.

If he wants it changed like that I'll push another amendmend to do that.

shumow commented:
Ok, thank you for the updates.  I'm fine with this, I'll give Marc a little bit more time to take a look at this.

cr-marcstevens commented:
Thanks! This does seem a good way to make it easy to include in other projects.
